contributors/design-proposals/cluster-lifecycle/draft-20171020-bootstrap-checkpointing.md

```diff
@@ -0,0 +1,108 @@
+title: Kubernetes Bootstrap Checkpointing Proposal
+authors:
+  - name: Timothy St. Clair
+    github: @timothysc
+    email: timothysc@gmail.com
+owning-sig: sig-cluster-lifecycle 
+participating-sigs:
+  - sig-node
+reviewers:
+  - @yujuhong 
+  - @luxas 
+  - @roberthbailey
+approvers:
+  - @yujuhong
+  - @roberbailey
+editor:
+  name: @timothysc
+creation-date: 2017-10-20
+last-updated: 2017-10-20
+status: draft
+```
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Objectives](#objectives)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [User Stories](#user-stories)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Unresolved Questions](#unresolved-questions)
+
+## Summary
+
+There are several methods to deploy a kubernetes cluster, one method that offers some unique advantages is self hosting.  The purpose of this proposal is to outline a method to checkpoint specific annotated pods, namely the control plane components, for the purpose of enabling self hosting.  
+
+The details of self hosting are beyond the scope of this proposal, and are outlined in the references listed below: 
+
+  - [Self Hosted Kubernetes][0]
+  - [Kubeadm Self Hosting][1]
+
+Extra details on this proposal, and its history, can be found in the links below: 
+
+  - [Bootstrap Checkpointing Draft 1][2]
+  - [Bootstrap Checkpointing Draft 2][3]
+  - [WIP Implementation][4]
+
+## Objectives
+
+The scope of this proposal is **bounded**, but has the potential for broader reuse in the future.  The reader should be mindful of the explicitly stated [Non-Goals](#non-goals) that are listed below.
+
+### Goals
+ 
+ - Provide a basic framework for recording annotated *Pods* to the filesystem. 
+ - Ensure that a restart of the kubelet checks for existence of these files and loads them on startup.
+
+### Non-Goals
+
+- This is not a generic checkpointing mechanism for arbitrary resources. (e.g. Secrets)  Such changes require wider discussions.
+- This will not checkpoint internal kubelet state. 
+- This proposal does not cover self hosted kubelet(s).  It is beyond the scope of this proposal, and comes with it's own unique set of challenges.
+
+## Proposal
+The enablement of this feature is gated by a single command line flag that is passed to the kubelet on startup, ```--bootstrap-checkpoint-path``` , and will be denoted that it is experimental. 
+
+### User Stories
+
+#### Pod Submission to Running
+- On submission of a Pod, via kubeadm or an operator, an annotation ```node.kubernetes.io/checkpoint=true``` is added to that Pod, which indicates that it should be checkpointed by the kubelet.  When the kubelet receives a notification from the apiserver that a new pod is to run, it will inspect the ```--bootstrap-checkpoint-path``` variable to determine if checkpointing is enabled.  Finally, the kubelet will perform an atomic write of a ```Pod_UID.yaml``` file when the annotation exists.  
```

Comment:
This sounds like all pods will have this annotation. Maybe say something like "... where an annotation blah has been added to that Pod, this indicates that it should be checkpointed by the kubelet."